modules/infra/control/loopback.c

```diff
@@ -0,0 +1,212 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2024 Christophe Fontaine
+
+#include <gr.h>
+#include <gr_control_input.h>
+#include <gr_control_output.h>
+#include <gr_eth.h>
+#include <gr_iface.h>
+#include <gr_infra.h>
+#include <gr_log.h>
+#include <gr_loopback.h>
+#include <gr_mempool.h>
+#include <gr_module.h>
+
+#include <event2/event.h>
+#include <rte_errno.h>
+#include <rte_malloc.h>
+
+#include <fcntl.h>
+#include <linux/if_tun.h>
+#include <net/if.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#define TUN_TAP_DEV_PATH "/dev/net/tun"
+
+static struct rte_mempool *loopback_pool;
+static struct event_base *global_ev_base;
+
+struct iface_info_loop {
+	int fd;
+	struct event *ev;
+};
+
+void loopback_tx(struct rte_mbuf *m) {
+	struct mbuf_data *d = mbuf_data(m);
+	struct iface_info_loop *lo;
+	void *data;
+
+	lo = (struct iface_info_loop *)d->iface->info;
+	if (rte_pktmbuf_linearize(m) == 0) {
+		data = rte_pktmbuf_mtod(m, char *);
+	} else {
+		data = rte_malloc(NULL, rte_pktmbuf_pkt_len(m), 0);
+		if (data == NULL) {
+			LOG(ERR, "rte_malloc failed %s", rte_strerror(rte_errno));
+			return;
+		}
+		data = (void *)rte_pktmbuf_read(m, 0, rte_pktmbuf_pkt_len(m), data);
+	}
+
+	if (write(lo->fd, data, rte_pktmbuf_pkt_len(m)) < 0)
+		LOG(ERR, "write to tun device %s failed %s", d->iface->name, strerror(errno));
+
+	if (!rte_pktmbuf_is_contiguous(m))
+		rte_free(data);
+
+	rte_pktmbuf_free(m);
+}
+
+static void iface_loopback_poll(evutil_socket_t, short, void *ev_iface) {
+	struct eth_input_mbuf_data *e;
+	struct iface_info_loop *lo;
+	struct iface *iface = ev_iface;
+	struct rte_mbuf *mbuf;
+	size_t len;
+	char *data;
+
+	lo = (struct iface_info_loop *)iface->info;
+
+	mbuf = rte_pktmbuf_alloc(loopback_pool);
+	if (!mbuf)
+		goto cleanup;
+
+	if (iface->mtu == 0)
+		iface->mtu = 1500;
+
+	if ((data = rte_pktmbuf_append(mbuf, iface->mtu)) == NULL)
+		goto cleanup;
+
+	if ((len = read(lo->fd, data, iface->mtu)) <= 0)
+		goto cleanup;
```

Comment:
Check in `main/api.c` for other potential errors returned by read() in a libevent callback.

Reply:
lib event wouldn't call this callback if there is no data to read, would it ?
I'm reworking a bit this part of code to use more common paradigms of libevent though.